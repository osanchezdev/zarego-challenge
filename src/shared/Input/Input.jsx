import React from 'react';
import { string, oneOf, bool, func } from 'prop-types';
import styles from './Input.module.css';
import searchIcon from '../../assets/icons/search.svg';

const Input = ({
  type,
  name,
  controller,
  placeholder,
  label,
  showLabel,
  icon,
  prefixText,
  error,
  errorMsg,
  onChangeParam,
  onFocusParam,
}) => {
  const { ref, onChange, onBlur } = controller(name);

  const handleChange = (e) => {
    onChange(e);
    onChangeParam?.(e);
  };

  return (
    <div className={styles.input__main_wrapper}>
      {label && showLabel && (
        <div className={styles.input_label__wrapper}>
          <label htmlFor={name}>{label}</label>
        </div>
      )}
      <div className={styles.input__wrapper}>
        {prefixText && (
          <span className={styles.input__prefix_text}>{prefixText}</span>
        )}
        <input
          className={`${styles.input} ${error ? styles.input_error : ''} ${
            icon ? styles.input_icon : ''
          }`}
          type={type}
          id={name}
          name={name}
          ref={ref}
          placeholder={placeholder}
          onChange={handleChange}
          onBlur={onBlur}
          onFocus={onFocusParam}
        />
        {icon && (
          <img
            src={searchIcon}
            alt="input-icon"
            className={styles.input_icon_image}
          />
        )}
        <div
          className={`${styles.input__error_msg_wrapper} ${
            prefixText ? styles.input__with_prefix_text : ''
          }`}
        >
          {error && <p className={styles.input__error_msg}>{errorMsg}</p>}
        </div>
      </div>
    </div>
  );
};

Input.propTypes = {
  type: oneOf(['text', 'number']),
  name: string.isRequired,
  controller: func.isRequired,
  placeholder: string,
  label: string.isRequired,
  showLabel: bool,
  icon: oneOf(['search']),
  prefixText: string,
  error: bool,
  errorMsg: string,
  onChangeParam: func,
  onFocusParam: func,
};

Input.defaultProps = {
  type: 'text',
  placeholder: '',
  showLabel: true,
  icon: null,
  prefixText: null,
  error: false,
  errorMsg: null,
  onChangeParam: null,
  onFocusParam: null,
};

export default Input;
