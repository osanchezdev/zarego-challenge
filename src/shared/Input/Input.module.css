.input__main_wrapper {
  position: relative;
}

.input_label__wrapper {
  position: relative;
  margin-bottom: 20px;
}

.input__wrapper {
  position: relative;
  display: flex;
  flex-wrap: wrap;
}

.input__prefix_text {
  flex: 0 0 30px;
  height: 38px;
  font-size: 28px;
  text-align: center;
}

.input {
  flex: 1 1 auto;
  height: 30px;
}

.input:focus,
.input:active,
.input:focus:active {
  outline: none;
}

.input.input_error {
  border: 2px inset #dc3545;
}

.input_icon_image {
  background: transparent;
  position: absolute;
  max-width: 26px;
  padding: 5px;
}

.input.input_icon {
  padding-left: 35px;
}

.input__error_msg_wrapper {
  position: relative;
  padding: 5px 0;
  flex: 0 0 100%;
}

.input__error_msg {
  position: relative;
  margin: 0;
  color: #dc3545;
}
