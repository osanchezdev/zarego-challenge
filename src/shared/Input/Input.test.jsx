import React from 'react';
import { render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import renderer from 'react-test-renderer';
import Input from './Input';

// TODO: refactor
describe('Input component', () => {
  const controller = () => ({
    ref: jest.fn(),
    onChange: jest.fn(),
    onBlur: jest.fn(),
  });
  const { labelText, name, errorMsg, placeholder, text, prefix, icon } = {
    labelText: 'labelTest',
    name: 'nameField',
    errorMsg: 'Value is required',
    placeholder: 'Example',
    text: 'input test',
    prefix: '$',
    icon: 'search',
  };

  it('should render component', () => {
    const tree = renderer
      .create(<Input label={labelText} name={name} controller={controller} />)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('should render input label', () => {
    render(<Input label={labelText} name={name} controller={controller} />);
    const label = screen.getByLabelText(labelText);
    expect(label).toBeInTheDocument();
  });

  it('should have input placeholder', () => {
    render(
      <Input
        label={labelText}
        name={name}
        placeholder={placeholder}
        controller={controller}
      />
    );
    const input = screen.getByPlaceholderText(placeholder);
    expect(input).toBeInTheDocument();
  });

  it('should not render input label', () => {
    render(
      <Input
        label={labelText}
        name={name}
        showLabel={false}
        controller={controller}
      />
    );
    const label = screen.queryByLabelText(labelText);
    expect(label).not.toBeInTheDocument();
  });

  it('should render input error message', () => {
    render(
      <Input
        label={labelText}
        name={name}
        error
        errorMsg={errorMsg}
        controller={controller}
      />
    );
    const error = screen.getByText(errorMsg);
    expect(error).toBeInTheDocument();
  });

  it('should not render input error message', () => {
    render(
      <Input
        label={labelText}
        name={name}
        errorMsg={errorMsg}
        controller={controller}
      />
    );
    const error = screen.queryByText(errorMsg);
    expect(error).not.toBeInTheDocument();
  });

  it('should render input prefix text', () => {
    render(
      <Input
        label={labelText}
        name={name}
        prefixText={prefix}
        controller={controller}
      />
    );
    const prefixText = screen.getByText(prefix);
    expect(prefixText).toBeInTheDocument();
  });

  it('should not render input preffix text', () => {
    render(<Input label={labelText} name={name} controller={controller} />);
    const prefixText = screen.queryByText(prefix);
    expect(prefixText).not.toBeInTheDocument();
  });

  it('should render input icon', () => {
    render(
      <Input
        label={labelText}
        name={name}
        icon={icon}
        controller={controller}
      />
    );
    const iconImg = screen.getByAltText('input-icon');
    expect(iconImg).toBeInTheDocument();
  });

  it('should not render input icon', () => {
    render(<Input label={labelText} name={name} controller={controller} />);
    const iconImg = screen.queryByAltText('input-icon');
    expect(iconImg).not.toBeInTheDocument();
  });

  it('should have typed value', () => {
    render(
      <Input
        label={labelText}
        name={name}
        placeholder={placeholder}
        controller={controller}
      />
    );
    const input = screen.getByPlaceholderText(placeholder);

    userEvent.type(input, text);
    expect(input).toHaveValue(text);
  });
});
