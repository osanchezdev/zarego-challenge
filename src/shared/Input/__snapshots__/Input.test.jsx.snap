// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Input component should render component 1`] = `
<div
  className="input__main_wrapper"
>
  <div
    className="input_label__wrapper"
  >
    <label
      htmlFor="nameField"
    >
      labelTest
    </label>
  </div>
  <div
    className="input__wrapper"
  >
    <input
      className="input  "
      id="nameField"
      name="nameField"
      onBlur={[MockFunction]}
      onChange={[Function]}
      onFocus={null}
      placeholder=""
      type="text"
    />
    <div
      className="input__error_msg_wrapper "
    />
  </div>
</div>
`;
